File: backend/express/controllers/notificationController.js
const { queryGet, queryAll, queryRun } = require("../services/dbService");

exports.getNotifications = async (req, res) => {
  const { user_id } = req.user;

  try {
    const list = await queryAll(
      "SELECT * FROM notifications WHERE user_id = ? ORDER BY created_at DESC",
      [user_id]
    );
    res.json({ success: true, data: list });
  } catch (err) {
    res.status(500).json({ success: false, error: { code: 500, message: "Internal server error" } });
  }
};

exports.postNotification = async (req, res) => {
  const { userId, title, message, channel = "InApp" } = req.body;

  if (!userId || !title || !message) {
    return res.status(400).json({
      success: false,
      error: { code: 400, message: "userId, title, and message are required" }
    });
  }

  const now = new Date().toISOString();

  try {
    const result = await queryRun(
      "INSERT INTO notifications (user_id, title, message, status, channel, created_at) VALUES (?, ?, ?, 'unread', ?, ?)",
      [Number(userId), title, message, channel, now]
    );

    res.status(201).json({
      success: true,
      data: {
        id: result.lastID,
        user_id: Number(userId),
        title,
        message,
        status: "unread",
        channel,
        created_at: now
      }
    });
  } catch (err) {
    res.status(500).json({ success: false, error: { code: 500, message: "Internal server error" } });
  }
};

File: backend/flask/controllers/notification_controller.py
from flask import request, jsonify, g
from services.notification_service import NotificationService

def get_notifications_controller():
    # Load for current authenticated user
    user_id = g.user["user_id"]
    notifications = NotificationService.get_notifications(user_id)
    return jsonify({"success": True, "data": notifications}), 200

def post_notifications_controller():
    data = request.get_json(silent=True) or {}
    user_id = data.get("userId")
    title = data.get("title")
    message = data.get("message")
    channel = data.get("channel", "InApp")
    
    if not user_id or not title or not message:
        return jsonify({"success": False, "error": {"code": 400, "message": "userId, title, and message are required"}}), 400
        
    res = NotificationService.create_notification(int(user_id), title, message, channel)
    return jsonify({"success": True, "data": res}), 201
